File: src/core/operations/repos.ts
/**
 * 仓库相关的 Gitea API 操作。所有 UI 与 AI 工具均通过本模块访问仓库数据。
 */
import { GiteaApiError } from '../errors';
import type { GiteaClient } from '../giteaClient';
import { decodeBase64, encodeBase64, encodeRepoPath, looksLikeBase64 } from '../encoding';
import type {
  GiteaBranch,
  GiteaCombinedStatus,
  GiteaCommit,
  GiteaContentsResponse,
  GiteaFileResponse,
  GiteaListResult,
  GiteaRepository,
} from '../types';

/** 仓库列表查询参数。 */
export interface ListReposOptions {
  /** 指定组织或用户名；`mine=true` 时忽略。 */
  owner?: string;
  /** 关键词搜索。设置后走 `/repos/search`。 */
  search?: string;
  /** 仅列出当前用户拥有的仓库。 */
  mine?: boolean;
  /** 页码，从 1 开始。 */
  page?: number;
  /** 每页数量。 */
  limit?: number;
  /** 排序字段。 */
  sort?: 'alpha' | 'created' | 'updated' | 'size' | 'id';
  /** 排序方向。 */
  order?: 'asc' | 'desc';
}

/** 创建仓库参数。 */
export interface CreateRepoOptions {
  /** 仓库名（必填）。 */
  name: string;
  /** 所属组织；为空时创建到当前用户名下。 */
  owner?: string;
  /** 仓库描述。 */
  description?: string;
  /** 是否私有。 */
  private?: boolean;
  /** 是否自动初始化（创建 README）。 */
  autoInit?: boolean;
  /** 默认分支名。 */
  defaultBranch?: string;
  /** README 模板，例如 `Default`。 */
  readme?: string;
  /** .gitignore 模板名。 */
  gitignores?: string;
  /** 许可证模板名。 */
  license?: string;
  /** 是否为模板仓库。 */
  template?: boolean;
}

/** 列出提交的参数。 */
export interface ListCommitsOptions {
  /** 分支 / 标签 / 提交 SHA。 */
  ref?: string;
  /** 仅列出影响该路径的提交。 */
  path?: string;
  /** 是否返回文件变更明细。 */
  withFiles?: boolean;
  page?: number;
  limit?: number;
}

/** 读取文件内容的返回结构。 */
export interface RepoFileContent {
  /** 仓库内路径。 */
  path: string;
  /** 文件 blob SHA（更新文件时必需）。 */
  sha: string;
  /** 解码后的文本内容。 */
  content: string;
  /** 文件大小（字节）。 */
  size: number;
  /** 内容是否被截断。 */
  truncated: boolean;
}

/** 提交文件（创建或更新）的参数。 */
export interface CommitFileOptions {
  /** 文件内容（纯文本，内部会自动 base64）。 */
  content: string;
  /** 提交信息。 */
  message?: string;
  /** 目标分支。 */
  branch?: string;
  /** 设定后会在该新分支上提交。 */
  newBranch?: string;
  /** 已知的 blob SHA；提供则执行更新，否则先查询后决定。 */
  sha?: string;
  /** 作者信息。 */
  author?: { name: string; email: string };
}

/** 仓库操作集合。 */
export class RepoOperations {
  constructor(private readonly client: GiteaClient) {}

  /**
   * 列出仓库。
   * @param options 查询参数
   * @returns 仓库列表与分页信息
   */
  public async list(options: ListReposOptions = {}): Promise<GiteaListResult<GiteaRepository>> {
    const page = options.page ?? 1;
    const limit = options.limit ?? 50;

    if (options.search) {
      const result = await this.client.requestWithMeta<unknown>('GET', '/repos/search', {
        query: { q: options.search, page, limit, sort: options.sort, order: options.order },
      });
      return { items: unwrapSearchResults<GiteaRepository>(result.data), pageInfo: result.pageInfo };
    }

    const owner = options.owner?.trim();
    if (owner && !options.mine) {
      const path = await this.resolveOwnerReposPath(owner);
      const result = await this.client.requestWithMeta<GiteaRepository[]>('GET', path, {
        query: { page, limit },
      });
      return { items: result.data ?? [], pageInfo: result.pageInfo };
    }

    const result = await this.client.requestWithMeta<GiteaRepository[]>('GET', '/user/repos', {
      query: { page, limit },
    });
    return { items: result.data ?? [], pageInfo: result.pageInfo };
  }

  /**
   * 获取单个仓库。
   * @param owner 所属者
   * @param repo 仓库名
   * @returns 仓库详情
   */
  public get(owner: string, repo: string): Promise<GiteaRepository> {
    return this.client.request<GiteaRepository>('GET', `/repos/${enc(owner)}/${enc(repo)}`);
  }

  /**
   * 创建仓库。
   * @param options 创建参数
   * @returns 新建的仓库
   */
  public create(options: CreateRepoOptions): Promise<GiteaRepository> {
    const body: Record<string, unknown> = { name: options.name };
    if (options.description !== undefined) body.description = options.description;
    if (options.private !== undefined) body.private = options.private;
    if (options.autoInit !== undefined) body.auto_init = options.autoInit;
    if (options.defaultBranch !== undefined) body.default_branch = options.defaultBranch;
    if (options.readme !== undefined) body.readme = options.readme;
    if (options.gitignores !== undefined) body.gitignores = options.gitignores;
    if (options.license !== undefined) body.license = options.license;
    if (options.template !== undefined) body.template = options.template;

    const owner = options.owner?.trim();
    const path = owner ? `/orgs/${enc(owner)}/repos` : '/user/repos';
    return this.client.request<GiteaRepository>('POST', path, { body });
  }

  /**
   * 列出分支。
   * @param owner 所属者
   * @param repo 仓库名
   * @param page 页码
   * @param limit 每页数量
   * @returns 分支列表
   */
  public async listBranches(
    owner: string,
    repo: string,
    page = 1,
    limit = 50,
  ): Promise<GiteaListResult<GiteaBranch>> {
    const result = await this.client.requestWithMeta<GiteaBranch[]>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/branches`,
      { query: { page, limit } },
    );
    return { items: result.data ?? [], pageInfo: result.pageInfo };
  }

  /**
   * 创建分支。
   * @param owner 所属者
   * @param repo 仓库名
   * @param newBranch 新分支名
   * @param oldRef 源分支 / 标签 / 提交，缺省为仓库默认分支
   * @returns 新建的分支
   */
  public createBranch(
    owner: string,
    repo: string,
    newBranch: string,
    oldRef?: string,
  ): Promise<GiteaBranch> {
    const body: Record<string, unknown> = { new_branch_name: newBranch };
    if (oldRef) {
      body.old_ref_name = oldRef;
      body.old_branch_name = oldRef;
    }
    return this.client.request<GiteaBranch>('POST', `/repos/${enc(owner)}/${enc(repo)}/branches`, { body });
  }

  /**
   * 列出提交。
   * @param owner 所属者
   * @param repo 仓库名
   * @param options 查询参数
   * @returns 提交列表
   */
  public async listCommits(
    owner: string,
    repo: string,
    options: ListCommitsOptions = {},
  ): Promise<GiteaListResult<GiteaCommit>> {
    const result = await this.client.requestWithMeta<GiteaCommit[]>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/commits`,
      {
        query: {
          sha: options.ref,
          path: options.path,
          stat: options.withFiles ? 'true' : undefined,
          page: options.page ?? 1,
          limit: options.limit ?? 30,
        },
      },
    );
    return { items: result.data ?? [], pageInfo: result.pageInfo };
  }

  /**
   * 读取仓库内文件或目录。
   * @param owner 所属者
   * @param repo 仓库名
   * @param filePath 仓库内路径，空字符串表示根目录
   * @param ref 分支 / 标签 / 提交
   * @returns 目录条目数组或单个文件元数据
   */
  public getContents(
    owner: string,
    repo: string,
    filePath = '',
    ref?: string,
  ): Promise<GiteaContentsResponse | GiteaContentsResponse[]> {
    const path = encodeRepoPath(filePath);
    return this.client.request<GiteaContentsResponse | GiteaContentsResponse[]>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/contents${path ? `/${path}` : ''}`,
      { query: { ref } },
    );
  }

  /**
   * 读取文本文件内容，自动处理 base64 解码与大小截断。
   * @param owner 所属者
   * @param repo 仓库名
   * @param filePath 仓库内路径
   * @param ref 分支 / 标签 / 提交
   * @param maxBytes 最大读取字节数，默认 200KB
   * @returns 文件内容
   */
  public async getFileContent(
    owner: string,
    repo: string,
    filePath: string,
    ref?: string,
    maxBytes = 200_000,
  ): Promise<RepoFileContent> {
    const entry = await this.getContents(owner, repo, filePath, ref);
    if (Array.isArray(entry)) {
      throw new GiteaApiError(`路径 ${filePath} 是目录而非文件`, 400, 'GET contents', undefined);
    }
    const raw = entry.content ?? '';
    const content = entry.encoding === 'base64' || looksLikeBase64(raw) ? decodeBase64(raw) : raw;
    const truncated = content.length > maxBytes;
    return {
      path: entry.path,
      sha: entry.sha,
      content: truncated ? content.slice(0, maxBytes) : content,
      size: entry.size ?? Buffer.byteLength(content, 'utf8'),
      truncated,
    };
  }

  /**
   * 创建或更新文件并提交。
   * 未显式提供 `sha` 时会先查询目标文件是否存在，以决定走 POST（新建）还是 PUT（更新）。
   * @param owner 所属者
   * @param repo 仓库名
   * @param filePath 仓库内路径
   * @param options 提交参数
   * @returns 提交结果
   */
  public async commitFile(
    owner: string,
    repo: string,
    filePath: string,
    options: CommitFileOptions,
  ): Promise<GiteaFileResponse> {
    const path = encodeRepoPath(filePath);
    const endpoint = `/repos/${enc(owner)}/${enc(repo)}/contents/${path}`;
    const body: Record<string, unknown> = {
      content: encodeBase64(options.content),
      message: options.message ?? (options.sha ? `Update ${filePath}` : `Add ${filePath}`),
    };
    if (options.branch) body.branch = options.branch;
    if (options.newBranch) body.new_branch = options.newBranch;
    if (options.author) body.author = options.author;

    let sha = options.sha;
    if (sha === undefined) {
      sha = await this.findBlobSha(owner, repo, filePath, options.branch);
    }

    if (sha) {
      body.sha = sha;
      return this.client.request<GiteaFileResponse>('PUT', endpoint, { body });
    }
    return this.client.request<GiteaFileResponse>('POST', endpoint, { body });
  }

  /**
   * 获取某个提交 / 分支的合并状态（CI 汇总）。
   * @param owner 所属者
   * @param repo 仓库名
   * @param ref 分支 / 标签 / 提交 SHA
   * @returns 合并状态
   */
  public getCombinedStatus(owner: string, repo: string, ref: string): Promise<GiteaCombinedStatus> {
    return this.client.request<GiteaCombinedStatus>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/commits/${enc(ref)}/status`,
    );
  }

  /**
   * 查询文件当前的 blob SHA；文件不存在时返回 undefined。
   * @param owner 所属者
   * @param repo 仓库名
   * @param filePath 仓库内路径
   * @param ref 分支 / 标签 / 提交
   * @returns blob SHA 或 undefined
   */
  private async findBlobSha(
    owner: string,
    repo: string,
    filePath: string,
    ref?: string,
  ): Promise<string | undefined> {
    try {
      const entry = await this.getContents(owner, repo, filePath, ref);
      return Array.isArray(entry) ? undefined : entry.sha;
    } catch (error) {
      if (error instanceof GiteaApiError && error.isNotFound) {
        return undefined;
      }
      throw error;
    }
  }

  /**
   * 判断给定名称是组织还是用户，返回对应的仓库列表端点。
   * @param owner 组织或用户名
   * @returns API 路径
   */
  private async resolveOwnerReposPath(owner: string): Promise<string> {
    try {
      await this.client.request<unknown>('GET', `/orgs/${enc(owner)}`);
      return `/orgs/${enc(owner)}/repos`;
    } catch (error) {
      if (error instanceof GiteaApiError && error.isNotFound) {
        return `/users/${enc(owner)}/repos`;
      }
      throw error;
    }
  }
}

/** URL 路径片段编码。 */
function enc(value: string): string {
  return encodeURIComponent(value);
}

/**
 * 兼容 `/repos/search` 的两种响应形态：`{ok, data}` 与裸数组。
 * @param payload 原始响应
 * @returns 结果数组
 */
function unwrapSearchResults<T>(payload: unknown): T[] {
  if (Array.isArray(payload)) {
    return payload as T[];
  }
  if (payload && typeof payload === 'object' && Array.isArray((payload as { data?: unknown }).data)) {
    return (payload as { data: T[] }).data;
  }
  return [];
}

File: python/src/gitea_toolkit_mcp/operations/repos.py
"""仓库域操作（移植自 TS 版 `src/core/operations/repos.ts`）。"""

from __future__ import annotations

import base64
from typing import Any, Sequence
from urllib.parse import quote

from ..client import GiteaClient, ListResult

#: 搜索接口返回的是 ``{"data": [...]}`` 而不是裸数组，需要解包
def _unwrap_search(data: Any) -> Sequence[Any]:
    if isinstance(data, dict):
        return data.get("data") or []
    return data or []


def _enc(value: str) -> str:
    """路径段转义（仓库名与分支名可能含 ``/``）。"""
    return quote(str(value), safe="")


class RepoOperations:
    """仓库、分支、提交、文件。"""

    def __init__(self, client: GiteaClient) -> None:
        self._client = client

    # ── 仓库 ────────────────────────────────────────────────────

    async def list(
        self,
        *,
        owner: str | None = None,
        search: str | None = None,
        mine: bool = False,
        limit: int = 30,
    ) -> ListResult:
        """列出仓库。

        三个来源互斥（与 TS 版一致的优先级）：关键词搜索 → 指定 owner → 当前用户。
        """
        if search:
            return await self._client.request_paged(
                "GET",
                "/repos/search",
                limit=limit,
                query={"q": search},
                extract=_unwrap_search,
            )
        if owner:
            # owner 可能是用户也可能是组织，两个端点都试（TS 版同样处理）
            try:
                return await self._client.request_paged(
                    "GET",
                    f"/orgs/{_enc(owner)}/repos",
                    limit=limit,
                    extract=lambda data: data or [],
                )
            except Exception:
                return await self._client.request_paged(
                    "GET",
                    f"/users/{_enc(owner)}/repos",
                    limit=limit,
                    extract=lambda data: data or [],
                )
        if mine:
            return await self._client.request_paged(
                "GET",
                "/user/repos",
                limit=limit,
                extract=lambda data: data or [],
            )
        return await self._client.request_paged(
            "GET",
            "/repos/search",
            limit=limit,
            extract=_unwrap_search,
        )

    async def get(self, owner: str, repo: str) -> dict[str, Any]:
        """取仓库详情。"""
        return await self._client.request("GET", f"/repos/{_enc(owner)}/{_enc(repo)}")

    async def create(
        self,
        *,
        name: str,
        owner: str | None = None,
        description: str | None = None,
        private: bool = False,
        auto_init: bool = True,
        default_branch: str | None = None,
    ) -> dict[str, Any]:
        """创建仓库。给了 owner 就建到组织名下，否则建到当前用户名下。"""
        path = f"/orgs/{_enc(owner)}/repos" if owner else "/user/repos"
        body: dict[str, Any] = {
            "name": name,
            "private": private,
            "auto_init": auto_init,
        }
        if description:
            body["description"] = description
        if default_branch:
            body["default_branch"] = default_branch
        return await self._client.request("POST", path, body=body)

    # ── 分支 ────────────────────────────────────────────────────

    async def list_branches(self, owner: str, repo: str, *, limit: int = 50) -> ListResult:
        """列出分支。"""
        return await self._client.request_paged(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/branches",
            limit=limit,
            extract=lambda data: data or [],
        )

    async def create_branch(
        self,
        owner: str,
        repo: str,
        *,
        new_branch: str,
        from_ref: str | None = None,
    ) -> dict[str, Any]:
        """基于已有分支/标签/提交创建分支。省略 from 时用仓库默认分支。"""
        target = from_ref or await self.default_branch(owner, repo)
        return await self._client.request(
            "POST",
            f"/repos/{_enc(owner)}/{_enc(repo)}/branches",
            body={"new_branch_name": new_branch, "old_branch_name": target},
        )

    async def default_branch(self, owner: str, repo: str) -> str:
        """取仓库默认分支名。"""
        detail = await self.get(owner, repo)
        return detail.get("default_branch") or "main"

    # ── 提交 ────────────────────────────────────────────────────

    async def list_commits(
        self,
        owner: str,
        repo: str,
        *,
        ref: str | None = None,
        path: str | None = None,
        limit: int = 20,
    ) -> ListResult:
        """列出提交历史。"""
        query: dict[str, Any] = {}
        if ref:
            query["sha"] = ref
        if path:
            query["path"] = path
        return await self._client.request_paged(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/commits",
            limit=limit,
            query=query,
            # Gitea 在仓库为空时会返回 409/空体，统一按空列表处理
            extract=lambda data: data if isinstance(data, list) else [],
        )

    async def commit_status(self, owner: str, repo: str, ref: str) -> dict[str, Any]:
        """查询某个 ref 的 CI 合并状态。"""
        return await self._client.request(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/commits/{_enc(ref)}/status",
        )

    # ── 文件 ────────────────────────────────────────────────────

    async def list_files(
        self,
        owner: str,
        repo: str,
        *,
        path: str | None = None,
        ref: str | None = None,
    ) -> Any:
        """列出目录内容（返回数组）；Gitea 在目标是文件时会回单对象。"""
        suffix = f"/{quote(path, safe='/')}" if path else ""
        query: dict[str, Any] = {}
        if ref:
            query["ref"] = ref
        return await self._client.request(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/contents{suffix}",
            query=query,
        )

    async def get_file(
        self,
        owner: str,
        repo: str,
        *,
        file_path: str,
        ref: str | None = None,
    ) -> dict[str, Any]:
        """读取文件内容（自动 base64 解码），并保留 blob SHA 供后续更新。"""
        query: dict[str, Any] = {}
        if ref:
            query["ref"] = ref
        raw = await self._client.request(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/contents/{quote(file_path, safe='/')}",
            query=query,
        )
        if isinstance(raw, list):
            raise ValueError(f"{file_path} 是一个目录，请改用 gitea_list_files")
        encoded = raw.get("content") or ""
        try:
            decoded = base64.b64decode(encoded, validate=False).decode("utf-8")
        except (ValueError, UnicodeDecodeError):
            decoded = ""
        return {**raw, "decoded_content": decoded}

    async def commit_file(
        self,
        owner: str,
        repo: str,
        *,
        file_path: str,
        content: str,
        message: str | None = None,
        branch: str | None = None,
        new_branch: str | None = None,
    ) -> dict[str, Any]:
        """创建或更新文件并提交；文件已存在时自动带上 blob SHA。"""
        target_branch = branch or new_branch or await self.default_branch(owner, repo)
        path = f"/repos/{_enc(owner)}/{_enc(repo)}/contents/{quote(file_path, safe='/')}"

        # 已存在则取 sha（更新必须带）
        sha: str | None = None
        try:
            existing = await self._client.request(
                "GET", path, query={"ref": target_branch}
            )
            if isinstance(existing, dict):
                sha = existing.get("sha")
        except Exception:
            sha = None

        body: dict[str, Any] = {
            "content": base64.b64encode(content.encode("utf-8")).decode("ascii"),
            "message": message or f"chore: {'更新' if sha else '新增'} {file_path}",
            "branch": target_branch,
        }
        if sha:
            body["sha"] = sha
        if new_branch:
            body["new_branch"] = new_branch
        return await self._client.request("PUT", path, body=body)


__all__ = ["RepoOperations"]
